addons/sale/models/res_currency.py

```diff
@@ -0,0 +1,16 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import fields, models, _
+from odoo.exceptions import UserError
+
+
+class ResCurrency(models.Model):
+    _inherit = "res.currency"
+
+    def unlink(self):
+        if self.env["sale.order"].search([
+            ('currency_id', 'in', self.ids),
+        ], limit=1):
+            raise UserError(_("You cannot delete a currency used by some Sales Orders."))
+        return super().unlink()
```

Comment:
looks like all this should simply be `ondelete='restrict'` on `sale_order.currency_id`